User Auth Providers (#1745)
neo4j/docs-operations: modules/ROOT/pages/authentication-authorization/manage-users.adoc

[source, syntax, role="noheader"]
----
CREATE USER name [IF NOT EXISTS]
  [SET [PLAINTEXT \| ENCRYPTED] PASSWORD 'password']
  [[SET PASSWORD] CHANGE [NOT] REQUIRED]
  [SET STATUS {ACTIVE \| SUSPENDED}]
  [SET HOME DATABASE name]

>   [SET AUTH [PROVIDER] 'provider' "{"SET <key> <value>"}"]

Author: had a go at this. What do you reckon?

>   [SET AUTH [PROVIDER] 'provider' "{"SET <key> <value>"}"]

Reviewer: I think it's fine at least for now with the syntax version we have, then we'll hopefully get something easier after the discussions with docs and the other teams around adding syntax specifications in general to docs

  {[SET AUTH [PROVIDER] 'provider' "{"{SET <key> <value>}..."}"]}...
----

`<key><value>` pairs for the `SET AUTH` clause could include:
[source, syntax, role="noheader"]
----
SET AUTH [PROVIDER] 'provider' "{"
    {
        SET ID 'id' // a unique identifier of the user in an external system
        \| SET [PLAINTEXT \| ENCRYPTED] PASSWORD 'password'

>         \| SET PASSWORD CHANGE [NOT] REQUIRED

Author: added a comment in all relevant places

[NOTE]
====

> It is mandatory to specify a `SET PASSWORD` or a `SET AUTH` clause or both because users must have at least one auth provider.

Author: ```suggestion
It is mandatory to specify a `SET PASSWORD` and/or at least one `SET AUTH` clause because users must have at least one auth provider.
```

[NOTE]
====

> It is mandatory to specify a `SET PASSWORD` or a `SET AUTH` clause or both because users must have at least one auth provider.

Author: ```suggestion
It is mandatory to specify a `SET PASSWORD` and/or at least one `SET AUTH` clause because users must have at least one auth provider.
```

  [[SET PASSWORD] CHANGE [NOT] REQUIRED]
  [SET STATUS {ACTIVE \| SUSPENDED} ]
  [SET HOME DATABASE name]
  [SET AUTH [PROVIDER] 'provider' "{"SET <key> <value>"}"]
  [REMOVE HOME DATABASE]

>   [REMOVE { AUTH [PROVIDER[S]] provider[, ...] } \| { ALL AUTH [PROVIDER[S]] }]

Reviewer: this of course also applies below

----

> | Description

Reviewer: in the `Required privilege` part below there is a bit that is `[source, privilege, role="noheader"` and missing it's `]` which makes it look weird when rendered, while you are anyway touching this area it would be nice if you could add the `]` :)

Author: strangely, the markup does have the `]` but it is not rendering `[source, privilege, role="noheader"`.   Any idea whats wrong here @renetapopova ?

Reviewer: ```
[source, privilege, role="noheader"
----
GRANT SET USER STATUS
----
```
no, one of them is missing it

Author: ah yes, so it is. thanks. Done.

  [[SET PASSWORD] CHANGE [NOT] REQUIRED]
  [SET STATUS {ACTIVE \| SUSPENDED} ]
  [SET HOME DATABASE name]

>   {[SET AUTH [PROVIDER] 'provider' "{"{SET <key> <value>}..."}"]}...

Reviewer: not sure we need the `{}` around `[]` or if 
```suggestion
  [SET AUTH [PROVIDER] 'provider' "{"{SET <key> <value>}..."}"]...
```
is enough 🤔

  [REMOVE HOME DATABASE]

>   [REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]

Reviewer: the remove auth should also allow multiple
```suggestion
  [REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]...
```
or
```suggestion
  {[REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]}...
```
if we want the extra brackets that SET currently have

Reviewer: this of course still don't show you can have both REMOVE AUTH clauses but that might be ok as it won't give the user anything to add both 🤷

  [[SET PASSWORD] CHANGE [NOT] REQUIRED]
  [SET STATUS {ACTIVE \| SUSPENDED} ]
  [SET HOME DATABASE name]
  [SET AUTH [PROVIDER] 'provider' "{"{SET <key> <value>}..."}"]...
  [REMOVE HOME DATABASE]

>   [REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]...

Author: ah, yes. thanks. done

----
`<key><value>` pairs for the `SET AUTH` clause could include:
[source, syntax, role="noheader"]
----
SET AUTH [PROVIDER] 'provider' "{"
    {
        SET ID 'id' // a unique identifier of the user in an external system
        \| SET [PLAINTEXT \| ENCRYPTED] PASSWORD 'password' // PASSWORD clauses are only applicable to the 'native' provider
        \| SET PASSWORD CHANGE [NOT] REQUIRED // PASSWORD clauses are only applicable to the 'native' provider
    }
"}"
----

| Description
a|
Modifies the settings for an existing user.
At least one `SET` or `REMOVE` clause is required.

> `SET` and `REMOVE` clauses cannot be combined in the same command.

Author: done

| {check-mark}

| roles
| Native roles granted to the user using the `GRANT ROLE` command.

This is not necessarily the set of roles that a user will receive in practice. The latter depends on DMBS configuration as well as the user's xref:authentication-authorization/auth-providers.adoc[Auth Providers].

> Examples of where a user would receive a different set of roles from that which appears in this column would include cases where they use external (e.g. LDAP or OIDC) auth, or where they do not have the `native` xref:authentication-authorization/auth-providers.adoc[Auth Provider].

Reviewer: I think it's more if native isn't defined in config rather than if they have the object: as per previous discussion I believe you can have a user without a native auth object but still uses native roles for the authorization

Author: thanks. revised.

| {check-mark}

| roles
| Native roles granted to the user using the `GRANT ROLE` command.

This is not necessarily the set of roles that a user will receive in practice. The latter depends on DMBS configuration as well as the user's xref:authentication-authorization/auth-providers.adoc[Auth Providers].

Examples of where a user would receive a different set of roles from that which appears in this column would include where they use external (e.g. LDAP or OIDC) auth, or where `native` is not listed in the xref:configuration/configuration-settings.adoc#config_dbms.security.authorization_providers[`dbms.security.authorization_providers`] configuration setting.

Will return `null` in community edition.

> | LIST OF STRING

Reviewer: I know this isn't on you but shouldn't these have been updated to the Cypher types? so `LIST<STRING>` instead of `LIST OF STRING`? (would need to go over the whole table to check all of them conforms)

Reviewer: also applies to the SHOW CURRENT USER table

Reviewer: @renetapopova This seems to also apply to at least the different SHOW PRIVILEGE column tables as well, so it might be a slightly bigger task to just go over all of them and update where needed?

Reviewer 2: Hej @Hunterness, I am reviewing the PR now, but because it's enormous, affects many pages, and has hundreds of comments, I am doing it locally and might be looking only into some of the comments that you have discussed. I'll do this in one commit so you can see all my updates and comment on them if I have missed something.

Reviewer 2: By the way, I assume you have finished the review from your side, as I've already started updating it locally.

Reviewer: yeah I believe I was done (before I found this and the comment below as well when looking at another task XD)

Reviewer 2: Ok, then, when I am done, I'll look at all the unresolved comments to see if I can address something.

Reviewer 2: I think this should be addressed after we merge this one in a separate PR.

| {cross-mark}
| {check-mark}

| passwordChangeRequired

> | If `true`, the user must change their password at the next login. This will be null if the user has `native` auth disabled.

Reviewer: should these also be updated for the SHOW CURRENT USER table?

| {check-mark}

| roles
| Native roles granted to the user using the `GRANT ROLE` command.

The set of roles a user receives in practice may differ from those in this column.
It depends on DMBS configuration and the user's xref:authentication-authorization/auth-providers.adoc[auth providers].

For example, if they use external (e.g. LDAP or OIDC) auth, or if `native` is not listed in the xref:configuration/configuration-settings.adoc#config_dbms.security.authorization_providers[`dbms.security.authorization_providers`] configuration setting.

It returns `null` in Community edition.
| LIST OF STRING
| {cross-mark}
| {check-mark}

| passwordChangeRequired
| If `true`, the user must change their password at the next login.
This is `null` if the user has `native` auth disabled.
| BOOLEAN
| {check-mark}
| {check-mark}

| suspended
| If `true`, the user is currently suspended.

It returns `null` in Community edition.
| BOOLEAN
| {cross-mark}
| {check-mark}

| home

> | The home database configured by the user, otherwise `null`.

Author: Not related to this change, but...

```suggestion
| The home database configured for the user, otherwise `null`.
```

A home database is resolved if it is pointing to a database or a database alias.

> If this database is unavailable or no database is specified, the login will fail.

Reviewer 2: ```suggestion
If the configured home database is unavailable and the user does not specify another database, the login will fail.
```

|false
|false
|<null>
|"jake"
|["PUBLIC"]
|false
|false
|<null>
5+a|Rows: 2
|===

> To inspect users' xref:authentication-authorization/auth-providers.adoc[Auth Providers], use `SHOW USERS WITH AUTH`. This command will produce a row per user per Auth Provider and will yield the following two columns in addition to those output by `SHOW USERS`:

Reviewer: I think it might be better to have a sub-section of Listing Users which is Listing Users With Auth or something, that way we can refer to it directly and it is less likely to just disappear among the other things

(cause at the moment its not very visible what is examples/example descriptions/example results/tables of return columns for the command/info on behaviour)

Reviewer: If you want I could take a stab at restructuring this bit to be more readable/clear distinctions

Author: done

|===
|adminUser
|"neo4j"
1+a|Rows: 1
|===
======

[[access-control-list-user-auth-providers]]
=== Listing User Auth Providers

To inspect users' xref:authentication-authorization/auth-providers.adoc[Auth Providers], use `SHOW USERS WITH AUTH`. This command will produce a row per user per Auth Provider and will yield the following two columns in addition to those output by `SHOW USERS`:

[options="header", width="100%", cols="2a,4,2m,^.^,^.^"]
|===
| Column
| Description
| Type
| Community Edition
| Enterprise Edition

| provider
| The name of the auth provider.
| STRING
| {check-mark}
| {check-mark}

| auth
| A map containing configuration for the user. E.g. dn of the user for an `ldap` auth provider, the unique external identifier for an `oidc` auth provider, or password status for a native auth provider.

| MAP
| {check-mark}
| {check-mark}
|===

.Show users with auth
======
[source, cypher, role=noplay]
----
SHOW USERS WITH AUTH
----

.Result
[role="queryresult" options="header,footer", width="100%", cols="2m,3m,3m,2m,2m,3m,4m"]
|===
|user
|roles
|passwordChangeRequired
|suspended
|home
|provider
|auth
|"neo4j"
|["admin","PUBLIC"]
|false
|false
|<null>
|"native"
|{

> "password": "***",

Author: `*********` seemed to work.

> "password": "***",

Reviewer: amazing XD

"changeRequired": false
}
|"jack"
|["PUBLIC"]
|false
|false
|<null>
|"native"
|{
"password": "***",
"changeRequired": false
}
|"jack"
|["PUBLIC"]
|false
|false
|<null>
|"oidc1"
|{
"id": "jacksIdForOidc1"
}
7+a|Rows: 3
|===
======

.Show user with auth using filtering
======
Show all of the users with the `oidc` Auth Provider.
[source,cypher,role=noplay]
----
SHOW USERS WITH AUTH
WHERE provider = 'oidc1'
----

.Result
[role="queryresult" options="header,footer", width="100%", cols="2m,3m,3m,2m,2m,3m,4m"]
|===
|user
|roles
|passwordChangeRequired
|suspended
|home
|provider
|auth
|"jack"
|["PUBLIC"]
|false
|false
|<null>
|"oidc1"
|{
"id": "jacksIdForOidc1"
}
7+a|Rows: 1
|===

======

> [NOTE]

Reviewer: I think this note should move up to the main section and not listing auth section?

Author: to me this note says _"thats all for listing users. If you actually wanted to list their privileges, then do `SHOW USER name PRIVILEGES` etc."_

and so I could't find a place for it at one of the higher levels. Any ideas? 

Or we could re-word it to be a bit more descriptive?

Reviewer: since the with auth is a subsection this note now lies in the subsection so to me it gives the feel of it being part of the subsection and not the whole listing users in general (which is what it should be, it's not specific to the with auth part)?

Maybe just move it up to the syntax table (or expand what is there) and remove it from down here alltogether?

Author: when you mention _syntax table_, do you mean the `SHOW USER PRIVILEGES` syntax table on line 103? If so then the need for this note disappears? (there's nothing from the note to add to the `SHOW USER PRIVILEGES` syntax table which isn't already there, right?)

Reviewer: yeah, that one. I was more thinking if we wanted to switch 
```
For more information, see xref:authentication-authorization/manage-privileges.adoc#access-control-list-privileges[Listing privileges].
``` 
to something closer to 
```
The `SHOW USER name PRIVILEGES` command is described in xref:authentication-authorization/manage-privileges.adoc#access-control-list-privileges[Listing privileges].
```
from the note here, and then remove the note here

to be more explicit about it not being covered here at all and only over there

The `'password'` can either be a string value or a string parameter with default value length of at least 8 characters. +
The `PLAINTEXT` and `ENCRYPTED` keywords are optional and can be used to specify the format of the password, i.e. whether Neo4j needs to hash it or it has already been hashed.
By default, all passwords are encrypted (hashed) when stored in the Neo4j `system` database.
* The optional `PLAINTEXT` in `SET PLAINTEXT PASSWORD` has the same behavior as `SET PASSWORD`.
* The optional `ENCRYPTED` is used to recreate an existing user when the plaintext password is unknown, but the encrypted password is available in the _data/scripts/databasename/restore_metadata.cypher_ file of a database backup.
See xref:backup-restore/restore-backup#_restore_users_and_roles_metadata[Restore users and roles metadata]. +
With `ENCRYPTED`, the password string is expected to be in the format of `<encryption-version>,<hash>,<salt>`, where, for example:
** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.

> <3> Specifies whether the user must change their password at the next login.

Reviewer 2: ```suggestion
<3> Specifies whether the user must change their password at the next login. +
If the optional SET PASSWORD CHANGE [NOT] REQUIRED is omitted, the default is `CHANGE REQUIRED`. 
The SET PASSWORD part is only optional if it directly follows the SET PASSWORD clause.
```

> If omitted, the default is `CHANGE REQUIRED`.

Reviewer 2: ```suggestion
If the optional SET PASSWORD CHANGE [NOT] REQUIRED is omitted, the default is `CHANGE REQUIRED`. 
The SET PASSWORD part is only optional if it directly follows the SET PASSWORD clause.
```

The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.

> <4> The default for `SET STATUS` is `ACTIVE`.

Reviewer 2: ```suggestion
<4>  Specifies the user's status.
If not set, the default is `ACTIVE`.
```

SET HOME DATABASE anotherDb
----

> The equivalent command using the xref:authentication-authorization/auth-objects.adoc[Auth Objects] syntax would be:

Reviewer: should this one also have the `[role=label--new-5.24]`?

Author: yes, thanks. Added.

* The default for `SET STATUS` is `ACTIVE`.
* `SET HOME DATABASE` can be used to configure a home database for a user.
A home database will be resolved if it is either pointing to a database or a database alias.
If no home database is set, the DBMS default database is used as the home database for the user.
[role=label--new-5.24]
* One or more `SET AUTH` clauses can be used to set xref:authentication-authorization/auth-providers.adoc[Auth Providers], which define authentication / authorization providers for the user. This might be used to configure external auth providers like LDAP or OIDC, but can also be used as an alternative way to set the native (password-based) auth settings like `SET PASSWORD` and `SET PASSWORD CHANGE REQUIRED`.
Examples can be found below for `native`, xref:authentication-authorization/sso-integration.adoc#auth-sso-auth-providers[here] for OIDC and xref:authentication-authorization/ldap-integration.adoc#auth-ldap-auth-providers[here] for LDAP.
* It is mandatory to specify at least either a `SET PASSWORD` or a `SET AUTH` clause because users must have at least one Auth Provider.

> * The `SET PASSWORD CHANGE [NOT] REQUIRED`, `SET STATUS`, `SET AUTH`, and `SET HOME DATABASE` clauses can be applied in any order, as can their `REMOVE` counterparts where applicable. However, all `REMOVE` clauses must come before the first `SET` clause.

Author: thanks. fixed.

A home database will be resolved if it is either pointing to a database or a database alias.
If no home database is set, the DBMS default database is used as the home database for the user.

> * The `SET PASSWORD CHANGE [NOT] REQUIRED`, `SET STATUS`, and `SET HOME DATABASE` clauses can be applied in any order.

Author: done. moved both to last (after the `[role=label--new-5.24]`)

SET ENCRYPTED PASSWORD '1,6d57a5e0b3317055454e455f96c98c750c77fb371f3f0634a1b8ff2a55c5b825,190ae47c661e0668a0c8be8a21ff78a4a34cdf918cae3c407e907b73932bd16c' CHANGE NOT REQUIRED
SET STATUS ACTIVE
----

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

----
======

> .Show user with auth

Reviewer: separate from this PR as such but I do find it odd that we have two 'examples' above showing their results and then the parts that are actually rendered as examples don't have any results

Author: done

SET HOME DATABASE anotherDb
----

> The equivalent command using the xref:authentication-authorization/auth-providers.adoc[Auth Providers] syntax would be:

Reviewer: should this have the `[role=label--new-5.24]` as well?

Reviewer: looks like I do have 2 comments on this line XD

SET PLAINTEXT PASSWORD 'abcd1234'
----

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

SET ENCRYPTED PASSWORD '1,6d57a5e0b3317055454e455f96c98c750c77fb371f3f0634a1b8ff2a55c5b825,190ae47c661e0668a0c8be8a21ff78a4a34cdf918cae3c407e907b73932bd16c' CHANGE NOT REQUIRED
SET STATUS ACTIVE
----
[role=label--new-5.24]
The equivalent command using the xref:authentication-authorization/auth-providers.adoc[Auth Providers] syntax would be:

[source,cypher,role=noplay]
----
CREATE USER jake
SET STATUS ACTIVE
SET AUTH 'native' {
  SET ENCRYPTED PASSWORD '1,6d57a5e0b3317055454e455f96c98c750c77fb371f3f0634a1b8ff2a55c5b825,190ae47c661e0668a0c8be8a21ff78a4a34cdf918cae3c407e907b73932bd16c'
  SET PASSWORD CHANGE NOT REQUIRED
}
----

======

[NOTE, role=label--enterprise-edition]
====

> The `SET STATUS {ACTIVE | SUSPENDED}`, `SET HOME DATABASE` and `SET AUTH` parts of the commands are only available in Neo4j Enterprise Edition.

Author: Clarified, thanks.

This is equivalent to running `DROP USER jake IF EXISTS` followed by `CREATE USER jake SET PASSWORD 'abcd1234'`.

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

[NOTE, role=label--enterprise-edition]
====

> The `SET STATUS {ACTIVE | SUSPENDED}`, `SET HOME DATABASE` parts of the commands are only available in Neo4j Enterprise Edition. The `{SET | REMOVE} AUTH` clause for external providers is only available in Neo4j Enterprise Edition. `{SET | REMOVE} AUTH 'native'` can be used in Neo4j Community Edition.

Author: > also, CREATE don't have REMOVE

done

> remove native doesn't work in native, only set

do you mean _remove natifve doesn't work in Community Edition, only set_ ? If so, then done. If not, then I'm not sure what you mean 😄

======

[[access-control-create-users]]
== Creating users

You can create users using one of the following Cypher commands, depending on whether you want to create a new user or replace an existing one.
In both cases, you can specify the user's password, whether they must change it at the next login, their status, home database, and auth provider settings.
The `SET` clauses can be applied in any order.

> It is mandatory to specify a `SET PASSWORD` or a `SET AUTH` clause or both because users must have at least one auth provider.

Author: ```suggestion
It is mandatory to specify a `SET PASSWORD` and/or at least one `SET AUTH` clause because users must have at least one auth provider.
```

The `PLAINTEXT` and `ENCRYPTED` keywords are optional and can be used to specify the format of the password, i.e. whether Neo4j needs to hash it or it has already been hashed.
By default, all passwords are encrypted (hashed) when stored in the Neo4j `system` database.
* The optional `PLAINTEXT` in `SET PLAINTEXT PASSWORD` has the same behavior as `SET PASSWORD`.
* The optional `ENCRYPTED` is used to recreate an existing user when the plaintext password is unknown, but the encrypted password is available in the _data/scripts/databasename/restore_metadata.cypher_ file of a database backup.
See xref:backup-restore/restore-backup#_restore_users_and_roles_metadata[Restore users and roles metadata]. +
With `ENCRYPTED`, the password string is expected to be in the format of `<encryption-version>,<hash>,<salt>`, where, for example:
** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.

<3> Specifies whether the user must change their password at the next login.

> If omitted, the default is `CHANGE REQUIRED`.

Reviewer 2: `CHANGE REQUIRED` can only be used with `SET PASSWORD` is used. So, the users are either required to change it or not. I don't think we need the above details here.

> If omitted, the default is `CHANGE REQUIRED`.

Reviewer: it feels odd to me to say that it defaults when omitted when that is only true if you give native auth and not if you have neither, to me this says `if not here, it gets this value regardless of other clauses`. I don't think we have a sentence that says it isn't allowed at all unless you set the password?

> If omitted, the default is `CHANGE REQUIRED`.

Reviewer 2: `[[SET PASSWORD] CHANGE [NOT] REQUIRED] # <3>`
Ok, I was confused by the syntax. Now that I copied it I see that we have anothe pair of square brackets.

> If omitted, the default is `CHANGE REQUIRED`.

Reviewer 2: I'll revert it to what it was before.

  [[SET PASSWORD] CHANGE [NOT] REQUIRED]
  [SET STATUS {ACTIVE | SUSPENDED}]
  [SET HOME DATABASE name]
  {[SET AUTH [PROVIDER] 'provider' "{"{SET <key> <value>}..."}"]}...
  [REMOVE HOME DATABASE]

>   [REMOVE { AUTH [PROVIDER[S]] provider[, ...] } \| { ALL AUTH [PROVIDER[S]] }]

Reviewer: missed updating this one (as my suggestion was only given on the first one I assume), also added the repeating here
```suggestion
  [REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]...
```
or to have the extra brackets like SET
```suggestion
  {[REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]}...
```

*** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
*** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.
* If the optional `SET PASSWORD CHANGE [NOT] REQUIRED` is omitted, the default is `CHANGE REQUIRED`.
* The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.
* The default for `SET STATUS` is `ACTIVE`.
* `SET HOME DATABASE` can be used to configure a home database for a user.
A home database will be resolved if it is either pointing to a database or a database alias.
If no home database is set, the DBMS default database is used as the home database for the user.

> [role=label--new-5.24]

Reviewer: I don't think these labels in the list render?

Author: hmm, no, they don't do they. Any idea @renetapopova ? Thanks.

Reviewer 2: I'll take a look when I start reviewing the PR. Is it ready for review on my end?

* `SET HOME DATABASE` can be used to configure a home database for a user.
A home database will be resolved if it is either pointing to a database or a database alias.
If no home database is set, the DBMS default database is used as the home database for the user.

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

> * One or more `SET AUTH` clauses can be used to set xref:authentication-authorization/auth-objects.adoc[Auth Objects], which define authentication / authorization providers for the user. This might be used to configure external auth providers like LDAP or OIDC, but can also be used as an alternative way to set the native (password-based) auth settings like `SET PASSWORD` and `SET PASSWORD CHANGE REQUIRED`. TODO: Insert links to the LDAP and OIDC examples.

Reviewer: leftover TODO

Author: done, thanks

* `REMOVE HOME DATABASE` is used to unset the home database for a user.
This results in the DBMS default database being used as the home database for the user.

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

* `REMOVE AUTH` is used to remove one or many xref:authentication-authorization/auth-objects.adoc[Auth Object(s)] from a user. It cannot be used such that a user would have no Auth Objects.

> * `REMOVE ALL AUTH` is used to remove all existing xref:authentication-authorization/auth-objects.adoc[Auth Objects] from a user and replace them with new ones. It must be used in conjunction with at least one `SET AUTH` clause.

Reviewer: >from a user and replace them with new ones

make it sound like the remove clause adds new ones, might be better to just say removes them and that it is invalid without at least one SET AUTH to add new ones as replacements

Author: done, thanks

.Modify the user `bob` with a new password and active status, and remove the requirement to change his password:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET PASSWORD 'abcd5678' CHANGE NOT REQUIRED
SET STATUS ACTIVE
----

[role=label--new-5.24]
The equivalent command using the xref:authentication-authorization/auth-objects.adoc[Auth Objects] syntax would be:

----
ALTER USER bob
SET AUTH 'native' {SET PASSWORD 'abcd5678' SET PASSWORD CHANGE NOT REQUIRED}
SET STATUS ACTIVE
----
======

.Modify the user `bob` to expire his current password so that he must change it the next time he logs in:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET PASSWORD CHANGE REQUIRED
----

[role=label--new-5.24]
The equivalent command using the xref:authentication-authorization/auth-objects.adoc[Auth Objects] syntax would be:

----
ALTER USER bob
SET AUTH 'native' {SET PASSWORD CHANGE REQUIRED}
----
======

.Assign the user `bob` a different home database:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET HOME DATABASE anotherDbOrAlias
----

> Or remove the `HOME` database from the user `bob`:

Reviewer: no, we usually refer to it as `home database` not ``` `HOME` database ```

Reviewer: also this line didn't get the `. ======` refactoring but seems to have been added as part of the assign a new one example?

Author: ah yes, I had intended that it be part of the same example, but you're right, its better as its own example. Changed.

* `REMOVE HOME DATABASE` is used to unset the home database for a user.
This results in the DBMS default database being used as the home database for the user.
[role=label--new-5.24]
* `REMOVE AUTH` is used to remove one or many xref:authentication-authorization/auth-objects.adoc[Auth Object(s)] from a user. It cannot be used such that a user would have no Auth Objects.
* `REMOVE ALL AUTH` is used to remove all existing xref:authentication-authorization/auth-objects.adoc[Auth Objects] from a user and replace them with new ones. It must be used in conjunction with at least one `SET AUTH` clause.

.Modify the user `bob` with a new password and active status, and remove the requirement to change his password:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET PASSWORD 'abcd5678' CHANGE NOT REQUIRED
SET STATUS ACTIVE
----

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

The equivalent command using the xref:authentication-authorization/auth-objects.adoc[Auth Objects] syntax would be:

----
ALTER USER bob
SET AUTH 'native' {SET PASSWORD 'abcd5678' SET PASSWORD CHANGE NOT REQUIRED}
SET STATUS ACTIVE
----
======

.Modify the user `bob` to expire his current password so that he must change it the next time he logs in:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET PASSWORD CHANGE REQUIRED
----

> [role=label--new-5.24]

Reviewer: remember to update number to correct version once that is settled

  [[SET PASSWORD] CHANGE [NOT] REQUIRED]
  [SET STATUS {ACTIVE | SUSPENDED}]
  [SET HOME DATABASE name]
  {[SET AUTH [PROVIDER] 'provider' "{"{SET <key> <value>}..."}"]}...
  [REMOVE HOME DATABASE]
  [REMOVE { AUTH [PROVIDER[S]] provider[, ...] \| ALL AUTH [PROVIDER[S]] }]...
----
`<key><value>` pairs for the `SET AUTH` clause could include:
[source, syntax, role="noheader"]
----
SET AUTH [PROVIDER] 'provider' "{"
    {
        SET ID 'id' // a unique identifier of the user in an external system
        \| SET [PLAINTEXT \| ENCRYPTED] PASSWORD 'password' // PASSWORD clauses are only applicable to the 'native' provider
        \| SET PASSWORD CHANGE [NOT] REQUIRED // PASSWORD clauses are only applicable to the 'native' provider
    }
"}"
----

> * At least one `SET` or `REMOVE` clause is required for the command.

Author: done, apart from: 

> (the comments about order of clauses should also be updated to mention SET AUTH and how the order of the REMOVE clauses behave)

which bit are you referring to here?

> * At least one `SET` or `REMOVE` clause is required for the command.

Reviewer: that just like the SET clauses you can add the REMOVE clauses in any order, but they all need to be before the first SET clause

> * At least one `SET` or `REMOVE` clause is required for the command.

Author: ah, makes sense. Thanks. Amended.

[NOTE, role=label--enterprise-edition]
====

> The `SET STATUS {ACTIVE | SUSPENDED}`, `SET HOME DATABASE`, `REMOVE HOME DATABASE` parts of the command are only available in Neo4j Enterprise Edition.  The `{SET | REMOVE} AUTH` clause for external providers is only available in Neo4j Enterprise Edition. However, `SET AUTH 'native'` can be used in Neo4j Community Edition.

Reviewer: ```suggestion
The `SET STATUS {ACTIVE | SUSPENDED}`, `SET HOME DATABASE`, `REMOVE HOME DATABASE`, and `REMOVE AUTH` parts of the command are only available in Neo4j Enterprise Edition.  The `SET AUTH` clause for external providers is only available in Neo4j Enterprise Edition. However, `SET AUTH 'native'` can be used in Neo4j Community Edition.
```

[NOTE, role=label--enterprise-edition]
====

> The `SET STATUS {ACTIVE | SUSPENDED}`, `SET HOME DATABASE`, `REMOVE HOME DATABASE`, `SET AUTH` and `REMOVE AUTH` parts of the command are only available in Neo4j Enterprise Edition.

Author: updated in both places. thanks.

The `PLAINTEXT` and `ENCRYPTED` keywords are optional and can be used to specify the format of the password, i.e. whether Neo4j needs to hash it or it has already been hashed.
By default, all passwords are encrypted (hashed) when stored in the Neo4j `system` database.
* The optional `PLAINTEXT` in `SET PLAINTEXT PASSWORD` has the same behavior as `SET PASSWORD`.
* The optional `ENCRYPTED` is used to recreate an existing user when the plaintext password is unknown, but the encrypted password is available in the _data/scripts/databasename/restore_metadata.cypher_ file of a database backup.
See xref:backup-restore/restore-backup#_restore_users_and_roles_metadata[Restore users and roles metadata]. +
With `ENCRYPTED`, the password string is expected to be in the format of `<encryption-version>,<hash>,<salt>`, where, for example:
** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.
<5> Specifies whether the user must change their password at the next login.

> If the optional `SET PASSWORD CHANGE [NOT] REQUIRED` is omitted, the default is `CHANGE REQUIRED`.

Reviewer 2: ```suggestion
<5> Specifies whether the user must change their password at the next login.
If the optional `SET PASSWORD CHANGE [NOT] REQUIRED` is omitted when adding native auth to a user (either by first removing pre-existing native auth or if the user does not have native auth to start with), the default is `CHANGE REQUIRED`.
```

The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.

> <6> The default for `SET STATUS` is `ACTIVE`.

Reviewer 2: I prefer not to remove it as we explain each line of the syntax.

The `PLAINTEXT` and `ENCRYPTED` keywords are optional and can be used to specify the format of the password, i.e. whether Neo4j needs to hash it or it has already been hashed.
By default, all passwords are encrypted (hashed) when stored in the Neo4j `system` database.
* The optional `PLAINTEXT` in `SET PLAINTEXT PASSWORD` has the same behavior as `SET PASSWORD`.
* The optional `ENCRYPTED` is used to recreate an existing user when the plaintext password is unknown, but the encrypted password is available in the _data/scripts/databasename/restore_metadata.cypher_ file of a database backup.
See xref:backup-restore/restore-backup#_restore_users_and_roles_metadata[Restore users and roles metadata]. +
With `ENCRYPTED`, the password string is expected to be in the format of `<encryption-version>,<hash>,<salt>`, where, for example:
** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.
<5> Specifies whether the user must change their password at the next login.

> If omitted, the default is `CHANGE REQUIRED`.

Reviewer: for ALTER this is actually more nuanced.

* If the user has native auth already (previously always true), and no change require clause is given then the value of the flag isn't updated and the old value is kept.
* if the user doesn't have native auth already (so only has external auths) **and** a set password clause is given, then we default to `CHANGE REQUIRED` when this clause is omitted.
* if the user doesn't have native auth already (so only has external auths), and we have neither a set password nor a change required clause then it behaves as the create case and doesn't add the change required (defaults to null)
* lastly same as for create, if the user doesn't have native auth already (so only has external auths) and only the change required is given this is an error as we don't have a password

Reviewer 2: But these 2 sentences just want to explain what the syntax means with the native authentication as with create.

Reviewer: yeah, but it is wrong (and I think it was before as well). In the alter case we don't have add a default unless you add native auth to the user where it didn't have it before, otherwise we just keep what the user already have for this when omitted

The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.
<6> The default for `SET STATUS` is `ACTIVE`.
<7> Specifies a home database for a user. A home database is resolved if it is pointing to a database or a database alias. If no home database is set, the DBMS default database is used as the home database for that user.

<8> label:new[Introduced in 5.24] One or more `SET AUTH` clauses can be used to set xref:authentication-authorization/auth-providers.adoc[auth providers], which define authentication / authorization providers for that user.
This might be used to configure external auth providers like LDAP or OIDC, but can also be used as an alternative way to set the native (password-based) auth settings like `SET PASSWORD` and `SET PASSWORD CHANGE REQUIRED`.

> For examples, see <<access-control-create-users,Creating users>>, xref:authentication-authorization/sso-integration.adoc#auth-sso-auth-providers[Configure SSO at the user level using auth providers], and xref:authentication-authorization/ldap-integration.adoc#auth-ldap-auth-providers[Configure authentication/authorization at the user level using auth providers].

Reviewer: shouldn't this link down to alter user examples rather than create user examples for the native case?

Reviewer 2: I don't know. This is how it was.

Reviewer: might just have been me missing it in previous reviews then 🤷

[[access-control-alter-users]]
== Modifying users

You can modify users with the `ALTER USER` command.
The command allows you to change the user's password, status, home database, and auth provider settings.
The `SET` and `REMOVE` clauses can be applied in any order.
However, all `REMOVE` clauses must come before the first `SET` clause and at least one `SET` or `REMOVE` clause is required for the command.

> 

Reviewer 2: ```suggestion
If any of the `SET` or `REMOVE` clauses are omitted, the corresponding settings will not be changed.

```

.Modify the user `bob` with a new password and active status, and remove the requirement to change his password:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET PASSWORD 'abcd5678' CHANGE NOT REQUIRED
SET STATUS ACTIVE
----

[role=label--new-5.24]
The equivalent command using the xref:authentication-authorization/auth-providers.adoc[Auth Providers] syntax would be:

----
ALTER USER bob
SET AUTH 'native' {SET PASSWORD 'abcd5678' SET PASSWORD CHANGE NOT REQUIRED}
SET STATUS ACTIVE
----
======

.Modify the user `bob` to expire his current password so that he must change it the next time he logs in:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET PASSWORD CHANGE REQUIRED
----

[role=label--new-5.24]
The equivalent command using the xref:authentication-authorization/auth-providers.adoc[Auth Providers] syntax would be:

----
ALTER USER bob
SET AUTH 'native' {SET PASSWORD CHANGE REQUIRED}
----
======

.Assign the user `bob` a different home database:
======
[source,cypher,role=noplay]
----
ALTER USER bob
SET HOME DATABASE anotherDbOrAlias
----
======

> .Remove the home database from the user `bob`:

Author: done. there are some more in-depth Auth Providers examples  hyperlinked on linked on line 1020 for more detail, but agree that the three you mention are good to illustrate the way that `REMOVE` and `SET` work together.

** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.
<5> Specifies whether the user must change their password at the next login.
If the optional `SET PASSWORD CHANGE [NOT] REQUIRED` is omitted, the default is `CHANGE REQUIRED`.
The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.

> <6> The default for `SET STATUS` is `ACTIVE`.

Reviewer: Both of these seem wrong for ALTER as they imply things change even if not given. That is not how alter works.

It might help to add a sentence up around line 984 about only the specified parts of the ALTER command gets altered, if any clause is omitted the user keeps the current value of that part.

Then I'd probably just remove `<6>` and update `<5>` to say
```
<5> Specifies whether the user must change their password at the next login.
If the optional `SET PASSWORD CHANGE [NOT] REQUIRED` is omitted when adding native auth to a user (either by first removing pre-existing native auth or if the user does not have native auth to start with), the default is `CHANGE REQUIRED`.
The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.
```
just to actually capture the fact that we only have a default in one specific case

** `0` is the first version and refers to the `SHA-256` cryptographic hash function with iterations `1`.
** `1` is the second version and refers to the `SHA-256` cryptographic hash function with iterations `1024`.
<5> Specifies whether the user must change their password at the next login.
If the optional `SET PASSWORD CHANGE [NOT] REQUIRED` is omitted when adding native auth to a user (either by first removing pre-existing native auth or if the user does not have native auth to start with), the default is `CHANGE REQUIRED`.
The `SET PASSWORD` prefix of the `CHANGE [NOT] REQUIRED` clause is only optional if it directly follows the `SET PASSWORD 'password'` clause and is not part of a `SET AUTH` clause.

> <6> Specifies the user's status.
> The default is `ACTIVE`.

Reviewer: ```suggestion
<6> Specifies the user's status.
```

Reviewer 2: Ah, ok.
